File: 4/4.7.dart
//4.7~4.8

class AttackPower {
  static final _MIN = 0;
  // [BAD]
  // 定数になっていないのであとからどう変わるか分からない
  late int value;

  AttackPower(int value) {
    // [GOOD]
    // guard節で不正な値が代入されないようにしている
    if (value < _MIN) {
      throw Exception();
    }
    this.value = value;
  }
}

class Weapon {
  final AttackPower attackPower;
  // [GOOD]
  // インスタンス生成時に値が決まる定数のイニシャライザの書き方
  Weapon(AttackPower attackPower) : this.attackPower = attackPower {}
}

void main() {
  // [BAD]
  // クラスのインスタンスをパラメータとして渡して使い回している
  final attackPower = AttackPower(20);
  final weaponA = Weapon(attackPower);
  final weaponB = Weapon(attackPower);

  print("weapon a attack power : ${weaponA.attackPower.value}");
  print("weapon b attack power : ${weaponB.attackPower.value}");
}

File: 4/4.7.swift
class AttackPower {
    static let MIN = 0
    // [BAD]
    // 定数になっていないのであとからどう変わるか分からない
    var value: Int

    init(value: Int) throws {
        // [GOOD]
        // guard節で不正な値が代入されないようにしている
        guard value >= AttackPower.MIN else {
            throw fatalError()
        }
        self.value = value
    }
}

class Weapon {
    // [GOOD]
    // 定数パラメータ
    let attackPower: AttackPower

    init(attackPower: AttackPower) {
        self.attackPower = attackPower
    }
}

// [BAD]
// クラスのインスタンスをパラメータとして渡して使い回している
// [おまけ]
// 本当はtry!などせずcatchで処理すべきだが、話が逸れるため使っている
let attackPower = try! AttackPower(value: 20)
let weaponA = Weapon(attackPower: attackPower)
let weaponB = Weapon(attackPower: attackPower)

print("weapon a attack power : \(weaponA.attackPower.value)")
print("weapon b attack power : \(weaponB.attackPower.value)")
